opba-protocols/hbci-protocol/src/main/java/de/adorsys/opba/protocol/hbci/entrypoint/pis/HbciPrepareContext.java

```diff
@@ -0,0 +1,90 @@
+import org.mapstruct.MappingTarget;
+import org.mapstruct.NullValuePropertyMappingStrategy;
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+
+import static de.adorsys.opba.protocol.bpmnshared.GlobalConst.SPRING_KEYWORD;
+import static de.adorsys.opba.protocol.hbci.constant.GlobalConst.HBCI_MAPPERS_PACKAGE;
+
+@Component
+@RequiredArgsConstructor
+public class HbciPrepareContext {
+    private final HbciPrepareContext.FromRequest fromRequest;
+    private final HbciExtendWithServiceContext extender;
+    private final FlowableObjectMapper mapper;
+    private final HbciPaymentContextMergeMapper mergeContextMapper;
+    private final FlowableProperties properties;
+
+    @SneakyThrows
+    protected PaymentHbciContext prepareContext(ServiceContext<? extends FacadeServiceableGetter> serviceContext, ProtocolAction action) {
+        PaymentHbciContext context = fromRequest.map2Context(serviceContext.getRequest());
+        context.setAction(action);
+        extender.extend(context, serviceContext);
+
+        Bank bank = new Bank();
+        bank.setBic(serviceContext.getRequestScoped().aspspProfile().getBic());
+        bank.setBankCode(serviceContext.getRequestScoped().aspspProfile().getBankCode());
+        context.setBank(bank);
+
+        ProtocolFacingPayment payment = serviceContext.getRequestScoped().paymentAccess().getFirstByCurrentSession();
+        PaymentHbciContext savedPaymentContext = deserializePaymentHbciContext(payment);
+        mergeContextMapper.merge(context, savedPaymentContext);
+        savedPaymentContext.getPayment().setPaymentId(payment.getPaymentId());
+        savedPaymentContext.getHbciDialogConsent().setWithHktan(false);
+        return savedPaymentContext;
+    }
+
+    /**
+     * Mapper to convert incoming user request to processable request context.
+     */
+    @Mapper(componentModel = SPRING_KEYWORD, uses = HbciUuidMapper.class, implementationPackage = HBCI_MAPPERS_PACKAGE)
+    public interface FromRequest extends DtoMapper<InitiateSinglePaymentRequest, PaymentHbciContext> {
+
+        @Mapping(source = "facadeServiceable.bankId", target = "aspspId")
+        @Mapping(source = "facadeServiceable.requestId", target = "requestId")
+        @Mapping(source = "facadeServiceable.fintechRedirectUrlOk", target = "fintechRedirectUriOk")
+        @Mapping(source = "facadeServiceable.fintechRedirectUrlNok", target = "fintechRedirectUriNok")
+        PaymentHbciContext map2Context(FacadeServiceableGetter ctx);
+    }
+
+    @Mapper(componentModel = SPRING_KEYWORD, implementationPackage = HBCI_MAPPERS_PACKAGE,
+            nullValuePropertyMappingStrategy = NullValuePropertyMappingStrategy.IGNORE)
+    public interface HbciPaymentContextMergeMapper {
+        PaymentHbciContext merge(PaymentHbciContext source, @MappingTarget PaymentHbciContext target);
+    }
+
+    @SneakyThrows
+    private PaymentHbciContext deserializePaymentHbciContext(ProtocolFacingPayment target) {
```

Comment:
Can't we extract this function to some service class and use it from here and `HbciCachedResultAccessor` ?